src/bin/pg_autoctl/nodestate_utils.c

```diff
@@ -0,0 +1,374 @@
+ */
+void
+nodestatePrintNodeState(NodeAddressHeaders *headers,
+						CurrentNodeState *nodeState)
+{
+	char hostport[BUFSIZE] = { 0 };
+	char composedId[BUFSIZE] = { 0 };
+
+	(void) nodestatePrepareNode(headers,
+								&(nodeState->node),
+								nodeState->groupId,
+								hostport,
+								composedId);
+
+	fformat(stdout, "%*s | %*s | %*s | %*s | %*s | %*s | %*s\n",
+			headers->maxNameSize, nodeState->node.name,
+			headers->maxNodeSize, composedId,
+			headers->maxHostSize, hostport,
+			headers->maxLSNSize, nodeState->node.lsn,
+			headers->maxHealthSize, nodestateHealthToString(nodeState->health),
+			headers->maxStateSize, NodeStateToString(nodeState->reportedState),
+			headers->maxStateSize, NodeStateToString(nodeState->goalState));
+}
+
+
+/*
+ * nodestatePrepareNode prepares the "host:port" and the "Node" computed
+ * columns used to display a node. The hostport and composedId parameters must
+ * be pre-allocated string buffers.
+ */
+void
+nodestatePrepareNode(NodeAddressHeaders *headers, NodeAddress *node,
+					 int groupId, char *hostport, char *composedId)
+{
+	sformat(hostport, BUFSIZE, "%s:%d", node->host, node->port);
+
+	switch (headers->nodeKind)
+	{
+		case NODE_KIND_STANDALONE:
+		{
+			sformat(composedId, BUFSIZE, "%d", node->nodeId);
+			break;
+		}
+
+		default:
+		{
+			sformat(composedId, BUFSIZE, "%d/%d", groupId, node->nodeId);
+			break;
+		}
+	}
+}
+
+
+/*
+ * prepareHostNameSeparator fills in the pre-allocated given string with the
+ * expected amount of dashes to use as a separator line in our tabular output.
+ */
+void
+prepareHostNameSeparator(char nameSeparatorHeader[], int size)
+{
+	for (int i = 0; i <= size; i++)
+	{
+		if (i < size)
+		{
+			nameSeparatorHeader[i] = '-';
+		}
+		else
+		{
+			nameSeparatorHeader[i] = '\0';
+			break;
+		}
+	}
+}
+
+
+/*
+ * nodestateAsJSON populates a given JSON_Value with an JSON object that mimics
+ * the output from SELECT * FROM pgautofailover.current_state() by taking the
+ * information bits from the given nodeState.
+ */
+bool
+nodestateAsJSON(CurrentNodeState *nodeState, JSON_Value *js)
+{
+	JSON_Object *jsobj = json_value_get_object(js);
+
+	/* same field names as SELECT * FROM pgautofailover.current_state() */
+	json_object_set_number(jsobj, "node_id", (double) nodeState->node.nodeId);
+	json_object_set_number(jsobj, "group_id", (double) nodeState->groupId);
+	json_object_set_string(jsobj, "nodename", nodeState->node.name);
+	json_object_set_string(jsobj, "nodehost", nodeState->node.host);
+	json_object_set_number(jsobj, "nodeport", (double) nodeState->node.port);
+
+	json_object_set_string(jsobj, "current_group_state",
+						   NodeStateToString(nodeState->reportedState));
+
+	json_object_set_string(jsobj, "assigned_group_state",
+						   NodeStateToString(nodeState->goalState));
+
+	json_object_set_string(jsobj, "Minimum Recovery Ending LSN",
+						   nodeState->node.lsn);
+
+	json_object_set_number(jsobj, "health", (double) nodeState->health);
```

Comment:
Maybe nice to use `nodestateHealthToString` here as well.

Reply:
Done now.